lifecycle-watcher/src/main/java/org/apache/unomi/lifecycle/BundleWatcherImpl.java

```diff
@@ -223,50 +269,109 @@ private void displayLogsForInactiveServices() {
+        if (graphQLToInstall) {
+            featuresToInstall.add(CDP_GRAPHQL_FEATURE);
+            requiredBundlesFromFeatures.put("org.apache.unomi.cdp-graphql-api-impl", false);
+            requiredBundlesFromFeatures.put("org.apache.unomi.graphql-playground", false);
+        }
+    }
+
+    public boolean shouldInstallAdditionalFeatures() {
+        return !featuresToInstall.isEmpty();
+    }
+
+    private void installFeatures() {
+        List<String> installedFeatures = new ArrayList<>();
+        featuresToInstall.forEach(value -> {
+            try {
+                long featureStartupTime = System.currentTimeMillis();
+                if (!featuresService.isInstalled(featuresService.getFeature(value))) {
+                    System.out.println("Installing feature " + value);
+                    featuresService.installFeature(value, EnumSet.of(FeaturesService.Option.NoAutoRefreshManagedBundles,
+                            FeaturesService.Option.NoAutoRefreshUnmanagedBundles, FeaturesService.Option.NoAutoRefreshBundles));
+                    logger.info("Feature {} successfully installed in {} ms", value, System.currentTimeMillis() - featureStartupTime);
+                }
+                installedFeatures.add(value);
+            } catch (Exception e) {
+                logger.error("Error when installing {} feature", value, e);
+            }
+        });
+        installedFeatures.forEach(value -> featuresToInstall.remove(value));
+    }
+
+    private void startScheduler() {
+        if (scheduledFuture == null || scheduledFuture.isCancelled()) {
+            TimerTask task = new TimerTask() {
+                @Override
+                public void run() {
+                    displayLogsForInactiveBundles(requiredBundles);
+                    displayLogsForInactiveServices();
+                    checkStartupComplete();
+                }
+            };
+            scheduledFuture = scheduler
+                    .scheduleWithFixedDelay(task, checkStartupStateRefreshInterval, checkStartupStateRefreshInterval, TimeUnit.SECONDS);
+        }
+    }
+
+    private void startSchedulerForAdditionalBundles() {
+        if (scheduledFutureAdditionalBundles == null || scheduledFutureAdditionalBundles.isCancelled()) {
+            TimerTask task = new TimerTask() {
+                @Override
+                public void run() {
+                    if (shouldInstallAdditionalFeatures() && !installingFeatureStarted) {
+                        installingFeatureStarted = true;
+                        installFeatures();
+                    }
+                    displayLogsForInactiveBundles(requiredBundlesFromFeatures);
+                    checkStartupComplete();
+                }
+            };
+            scheduledFutureAdditionalBundles = scheduler
+                    .scheduleWithFixedDelay(task, checkStartupStateRefreshInterval, checkStartupStateRefreshInterval, TimeUnit.SECONDS);
+        }
+    }
```

Comment:
The code of this function could be reduce to the task it self being different. 
(In case we reuse the same ScheduledFuture variable like suggested below.)

Or other option, just use one function here and pass everything as parameter:

- installFeature: bool -> false first time, true second time
- inactiveBundlesToWatch: requiredBundles first time, requiredBundlesFromFeatures second time
- checkInactiveServices: bool -> true first time, false second time

Doing so we can:

- reuse existing ScheduledFuture variable, because both cannot be start in parallel, the feature task is started after the normal bundle task is ended.
- And we can have a generic task that is able to handle both cases depending on the parameter listed previously.